### PR TITLE
ENH: Add PKCS#12 digital signature with original PDF

### PR DESCRIPTION
**Pull request by cyy-2024**
Hello, I added a new sign file and modified the cil file. I also provided a p12 file of the test. You can run the following commands on the CMD：pdfly sign input.pdf --output signed_output.pdf --p12 certs.p12 --password "123456"。
屏幕截图 2024-11-10 084817
This results in a PDF that retains the original PDF and adds the signature text and signature information.
If you have any questions, please ask, thanks for giving me the opportunity!

### CODE REVIEW

File: .idea/.gitignore [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# Default ignored files
```
Comment (Lucas-C) [inline .idea/.gitignore:1]: Could you please avoid versioning configuration files specific to your [IDE](https://en.wikipedia.org/wiki/Integrated_development_environment)?

FLOSS developpers use many different tools to write code, and it would be preferable to keep IDE-specific configurations files unversioned.

You can however add the `.idea/` directory to [this repo root `.gitignore`](https://github.com/py-pdf/pdfly/blob/main/.gitignore) 🙂

File: certificate.pem [hunk 2]
```diff
@@ -0,0 +1,22 @@
+-----BEGIN CERTIFICATE-----
```
Comment (Lucas-C) [inline certificate.pem:1]: Could you please move the certificate files (`certificate.pem`, `certs.p12` & `private_key.pem`) to the `resources/` directory?

You could also create a `resources/README.md` file, and mention how those files were created.

File: pdfly/cli.py [hunk 3]
```diff
@@ -37,6 +38,29 @@ def version_callback(value: bool) -> None:
     ),
     rich_markup_mode="rich",  # Allows to pretty-print commands documentation
 )
+@entry_point.command(name="sign", help="Sign a PDF file using a PKCS12 (.p12) certificate.")
```
Comment (Lucas-C) [inline pdfly/cli.py:41]:
```suggestion
@entry_point.command(name="sign", help=pdfly.sign.__doc__)
```
As you can see for other subcommands, we usually provide a lengthy description there,
that is stored in the dedicated module top docstring.

File: pdfly/cli.py [hunk 4]
```diff
@@ -37,6 +38,29 @@ def version_callback(value: bool) -> None:
...
+def sign(
+    input_pdf: Path = typer.Argument(..., help="Input PDF file."),
```
Comment (Lucas-C) [inline pdfly/cli.py:43]:
```suggestion
    input_pdf: Annotated[
        Path,
        typer.Argument(
            exists=True,
            file_okay=True,
            readable=True,
            resolve_path=True,
            help="Input PDF file.",
        ),
    ],
```
As can be seen in other subcommands, `typer` provides those `Annoted` types than defines some basic checks on the input file(s).
They are handy, and are a good substitute to the "manual" check that you performed below: `if not p12 or not p12.exists(): ...`

File: pdfly/sign.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
+"""Sign a PDF file with a PKCS#12 certificate, retaining original formatting and adding a signature page."""
+
+from pathlib import Path
+from fpdf import FPDF
+import fitz  # PyMuPDF for precise page content extraction
```
Comment (Lucas-C) [inline pdfly/sign.py:5]: This dependency is a new dependency, that `pdfly` currently does not have.

Why introducing it?

The feature-request issue suggested to use `fpdf2`, that is already a dependency of `pdfly`, and also provide signing ability: https://github.com/py-pdf/pdfly/issues/71

File: pdfly/cli.py [hunk 6]
```diff
@@ -37,6 +38,29 @@ def version_callback(value: bool) -> None:
...
+    p12: Optional[Path] = typer.Option(None, "--p12", help="PKCS12 certificate file."),
+    password: Optional[str] = typer.Option(None, "--password", help="Password for the certificate."),
+) -> None:
+    """
+    Sign a PDF file using a PKCS12 (.p12) certificate.
+    """
+    print(f"input_pdf: {input_pdf}")
+    print(f"output_pdf: {output_pdf}")
+    print(f"p12: {p12}")
+    print(f"password: {password}")
```
Comment (Lucas-C) [inline pdfly/cli.py:54]: This is fine for debugging, but those `print()` calls will have to be removed in the final version of this PR 🙂

File: pdfly/sign.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+        page = doc[page_num]
+        rect = page.rect
+        pdf.add_page(orientation="P" if rect.width < rect.height else "L")
+        pdf.set_auto_page_break(False)
+        text = page.get_text("text")
+        pdf.set_xy(0, 0)
+        pdf.set_font("Arial", size=12)
+        pdf.multi_cell(0, 10, txt=text)
+
+    # Step 3: Add a signature page
```
Comment (Lucas-C) [inline pdfly/sign.py:33]: This step is not necessary 🙂

The `pdfly sign` subcommand should NOT alter the PDF content at all.

However we could check if the document already has a signature...

File: pdfly/sign.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    pdf.add_page()
+    pdf.set_xy(10, 10)
+    pdf.set_font("Arial", size=12)
+    pdf.cell(200, 10, txt="This document is signed with PKCS#12.", ln=True)
+
+    # Step 4: Digitally sign the PDF
+    pdf.sign_pkcs12(str(p12_path), password=password.encode("utf-8"))
+    pdf.output(str(output_pdf))
```
Comment (Lucas-C) [inline pdfly/sign.py:41]: I do not think that the conversion to string is necessary there

**Comment by Lucas-C** [comment]
Thank you for starting working on this subject, an welcome on board @cyy-2024! 🙂

First, I would warmly recommend to start initiating some unit tests quite early,
in the spirit of [test-driven_development](https://en.wikipedia.org/wiki/Test-driven_development).
This is a good habit to have when writing software, especially when unit tests are expected in the end for the piece of code you are writing.

You can start with 2 base unit tests:
* one simple unit test where every thing goes well, and you have a final assertion that checks that the signature of the forged PDF can be validated with `endesive`
* one failing test case, for example for the case where the `.p12` cannot be open due to a wrong password being provided

**Comment by cyy-2024** [comment]
I feel sorry，but, thanks for the guidance, please let me try again

File: pdfly/cli.py (hunk 3 above)
Comment (cyy-2024) [reply pdfly/cli.py:41]: Previously it was I didn't understand the role of three quotes, sorry, I corrected it.

File: pdfly/sign.py (hunk 5 above)
Comment (cyy-2024) [reply pdfly/sign.py:5]: Because I want to keep the original pdf content and add a signature, fpdf2 always needs to update an empty temporary pdf，i have no idea to slove it.   :(

**Comment by Lucas-C** [comment]
> I feel sorry，but, thanks for the guidance, please let me try again

Do not feel sorry! 🙂
You did well, and this 1st PR was better and made faster than many first timers on GitHub 👍

My feedbacks are there to let you learn the contribution process on this repository,
and to explain why we do things the way we do them 🙂

Take your time on this PR, I'm sure you can do it.

File: .idea/.gitignore (hunk 1 above)
Comment (Lucas-C) [reply .idea/.gitignore:1]: If you meant "directory" instead of "diagram" then yes! 🙂

Else I'm not sure about which diagram you are refering to 😄

File: pdfly/sign.py (hunk 5 above)
Comment (Lucas-C) [reply pdfly/sign.py:5]: > fpdf2 always needs to update an empty temporary pdf，i have no idea to slove it. :(

You could share the code that you used, and the details of the error you face, and I'll to help you solve it 🙂

File: .idea/.gitignore (hunk 1 above)
Comment (cyy-2024) [reply .idea/.gitignore:1]: i mean the picture i showed. hhh

File: pdfly/cli.py (hunk 4 above)
Comment (cyy-2024) [reply pdfly/cli.py:43]: it will show:"AttributeError: 'NoneType' object has no attribute 'isidentifier'"

Comment (Lucas-C) [reply pdfly/cli.py:43]: Sorry but I do not understand this comment 😅

Maybe you wanted to reply in the thread below regarding `fitz` & `fpdf2`?
If so, just providing the error is not enough for me to figure it out.
You will have to share the code you used.

**Comment by cyy-2024** [comment]
I'm sorry that the teacher recently arranged for me to develop a project, and it may take at least next week.TT

**Comment by Lucas-C** [comment]
> I'm sorry that the teacher recently arranged for me to develop a project, and it may take at least next week.TT

Alright, no problem!

This can wait, I just wanted to know if you still wanted to continue working on this 🙂

Good luck with your project!

**Comment by MartinThoma** [comment]
@cyy-2024 Hey, thank you for contributing! I just wanted to check-in: Do you need help with finishing this PR? Do you want to finish it yourself or would it be fine for you if somebody else took over (while still giving you credit in the commit message that this was your contribution, of course)

**Comment by Lucas-C** [comment]
> I am very sorry that my school assignment has not been completed yet(&nbsp;There's been some hiccups).&nbsp;I would be very grateful if someone on the issue wants to take over. Sorry for the inconvenience.

Alright, no worries.
Good luck with your studies 👍 🙂

**Comment by MartinThoma** [comment]
Thank you for letting us know - and thank you for giving it a shot :-) 

Best wishes also from my side :-)
